src/ctapipe/io/dl2_tables_preprocessing.py

```diff
@@ -0,0 +1,474 @@
+                        unit=u.dimensionless_unscaled,
+                        description="prediction of the classifier, defined between [0,1],"
+                        " where values close to 1 mean that the positive class"
+                        " (e.g. gamma in gamma-ray analysis) is more likely",
+                    ),
+                    Column(
+                        name="weight",
+                        unit=u.dimensionless_unscaled,
+                        description="Event weight",
+                    ),
+                ]
+            )
+
+        return QTable(
+            names=[col.name for col in schema],
+            dtype=[col.dtype for col in schema],
+            units=[col.unit for col in schema]
+            if any(col.unit for col in schema)
+            else None,
+            meta={},
+        )
+
+
+class DL2EventLoader(Component):
+    """
+    Component for loading events and simulation metadata, applying preselection and optional derived column logic.
+    """
+
+    classes = [DL2EventPreprocessor]
+
+    # User-selectable event reading function and kwargs
+    event_reader_function = Unicode(
+        default_value="read_subarray_events_chunked",
+        help=(
+            "Function of TableLoader used to read event chunks. "
+            "E.g., 'read_subarray_events_chunked' or 'read_telescope_events_chunked'."
+        ),
+    ).tag(config=True)
+
+    event_reader_kwargs = Dict(
+        default_value={},
+        help="Extra keyword arguments passed to the event reading function, e.g., {'path': '/dl2/event/telescope/Reconstructor'}",
+    ).tag(config=True)
+
+    def __init__(self, file: Path, target_spectrum: "Spectra", **kwargs):  # noqa: F821
+        from ..irf.spectra import SPECTRA
+
+        super().__init__(**kwargs)
+        self.epp = DL2EventPreprocessor(parent=self)
+        self.target_spectrum = SPECTRA[target_spectrum]
+        self.file = file
+
+    def load_preselected_events(
+        self, chunk_size: int, obs_time: u.Quantity
+    ) -> tuple[QTable, int, dict]:
+        """
+        Load and filter events from the file.
+
+        Parameters
+        ----------
+        chunk_size : int
+            Size of chunks to read from the file.
+        obs_time : Quantity
+            Observation time to scale weights.
+
+        Returns
+        -------
+        table : QTable
+            Filtered and processed event table.
+        n_raw_events : int
+            Number of events before selection.
+        meta : dict
+            Metadata dictionary with simulation info and input spectrum.
+        """
+
+        opts = dict(dl2=True, simulated=True, observation_info=True)
+
+        with TableLoader(self.file, parent=self, **opts) as load:
+            header = self.epp.make_empty_table()
+            sim_info, spectrum = self.get_simulation_information(load, obs_time)
+            meta = {"sim_info": sim_info, "spectrum": spectrum}
+            bits = [header]
```

Comment:
Does it need to be [header] if later you append header again? There must be a nicer way...